webpack/hashed-folder-copy-plugin/src/HashedFolderCopyPlugin.ts

```diff
@@ -41,103 +34,57 @@ interface IAcornNode<TExpression> {
-    // Dependency templates are looked up by their constructor instance, so this
-    // must be able to resolve the same `ConstDependency` module that the
-    // project is using for compilation.
-    let constDependencyModulePath: string | undefined;
-    const parentModulePath: string | undefined = module.parent?.parent?.path;
-    if (parentModulePath) {
-      try {
-        constDependencyModulePath = Import.resolveModule({
-          modulePath: WEBPACK_CONST_DEPENDENCY_MODULE_PATH,
-          baseFolderPath: parentModulePath
-        });
-      } catch (e) {
-        // Ignore
-      }
-    }
+export class HashedFolderCopyPlugin implements webpack.WebpackPluginInstance {
+  public apply(compiler: webpack.Compiler): void {
+    const webpack: typeof import('webpack') = compiler.webpack;
+    const { HashedFolderDependency, HashedFolderDependencyTemplate } =
+      getHashedFolderDependencyForWebpackInstance(webpack);
 
-    // Failing that, resolve relative to this module.
-    this._ConstDependency = require(constDependencyModulePath || WEBPACK_CONST_DEPENDENCY_MODULE_PATH);
-  }
+    compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation: webpack.Compilation) => {
+      compilation.dependencyTemplates.set(HashedFolderDependency, new HashedFolderDependencyTemplate());
+    });
 
-  public apply(compiler: webpack.Compiler): void {
-    // Casting here because VersionDetection refers to webpack 5 typings
-    let needToWarnAboutWebpack5: boolean = VersionDetection.isWebpack5(
-      compiler as unknown as Webpack5.Compiler
-    );
+    const hashedFolderDependencies: IHashedFolderDependency[] = [];
+
+    compiler.hooks.thisCompilation.tap(PLUGIN_NAME, (compilation: webpack.Compilation) => {
+      compilation.hooks.finishModules.tapPromise(PLUGIN_NAME, async () => {
+        const inputFileSystem: webpack.Compiler['inputFileSystem'] = compiler.inputFileSystem;
+        const notImplementedFunction: () => never = () => {
+          throw new Error('Not implemented');
+        };
+        const globFs: glob.FileSystemAdapter = {
+          lstat: inputFileSystem.lstat?.bind(inputFileSystem) ?? notImplementedFunction,
+          stat: inputFileSystem.stat?.bind(inputFileSystem) ?? notImplementedFunction,
+          lstatSync: notImplementedFunction,
+          statSync: notImplementedFunction,
+          readdir: inputFileSystem.readdir?.bind(inputFileSystem) ?? notImplementedFunction,
+          readdirSync: notImplementedFunction
+        } as unknown as glob.FileSystemAdapter; // The Webpack typings are wrong on `readdir`
+
+        await Async.forEachAsync(
+          hashedFolderDependencies,
+          async (hashedFolderDependency) => {
+            await hashedFolderDependency.processAssetsAsync(compilation, globFs);
+          },
+          { concurrency: 10 }
+        );
+      });
+    });
 
     compiler.hooks.normalModuleFactory.tap(PLUGIN_NAME, (normalModuleFactory) => {
```

Comment:
You can access the factory from the second argument to `compilation.hooks.thisCompilation` which has a property `normalModuleFactory`. That will let you scope the `hashedFolderDependencies` array to the compilation callback.